scalding-core/src/main/scala/com/twitter/scalding/typed/MultiJoinFunction.scala

```diff
@@ -0,0 +1,122 @@
+package com.twitter.scalding.typed
+
+import com.twitter.scalding.serialization.Externalizer
+import java.io.Serializable
+
+/**
+ * This is a weakly typed multi-way join function. By construction,
+ * it should be kept in sync with the types in a Seq[TypedPipe[(K, Any)]]
+ *
+ * a more sophisticated typing could use an HList of TypedPipe
+ * and another more advanced coding here to prove the types line up.
+ * However, this is somewhat easy to test and only exposed to
+ * those writing backends, so we are currently satisfied with the
+ * weak typing in this case
+ *
+ * We use Externalizer internally to independently serialize each function
+ * in the composition. This, in principle, should allow Externalizer
+ * to work better since different functions may be serializable with
+ * Kryo or Java, but currently Externalizer has to use java or kryo
+ * for the entire object.
+ */
+sealed abstract class MultiJoinFunction[A, +B] extends Serializable {
+  def inputSize: Int
+  def apply(key: A, leftMost: Iterator[Any], rightStreams: Seq[Iterable[Any]]): Iterator[B]
+}
+
+object MultiJoinFunction extends Serializable {
+  final case class Casting[A, B]() extends MultiJoinFunction[A, B] {
+    def inputSize = 1
+    def apply(k: A, iter: Iterator[Any], empties: Seq[Iterable[Any]]) = {
+      require(empties.isEmpty, "this join function should never be called with non-empty right-most")
+      iter.asInstanceOf[Iterator[B]]
+    }
+  }
+
+  final case class PairCachedRight[K, A, B, C](
+    left: MultiJoinFunction[K, A],
+    right: MultiJoinFunction[K, B],
+    @transient fn: (K, Iterator[A], Iterable[B]) => Iterator[C]) extends MultiJoinFunction[K, C] {
+
+    private[this] val fnEx = Externalizer(fn)
+
+    val inputSize = left.inputSize + right.inputSize
+    private[this] val leftSeqCount = left.inputSize - 1
+
+    def apply(key: K, leftMost: Iterator[Any], rightStreams: Seq[Iterable[Any]]): Iterator[C] = {
+      require(rightStreams.size == inputSize - 1, s"expected ${inputSize} inputSize, found ${rightStreams.size + 1}")
```

Comment:
probably good to add a comment describing some of the logic here around the input sizes

This is that the RHS is composed of N streams and we only cache the far right of those streams in this case?

Reply:
Ohhh its how many of the rightStreams are already covered in prior joins by the left hand multijoin function? so we need to feed them in there?

Reply:
It is just an extra check (which should never possibly fail unless we have a programming bug) that the number of streams we are joining matches the the total joining operation we have.

Since we have one stream in leftMost, the others should be in rightStreams. I'll add this as a comment.